hector_pointcloud_decimator/src/pointcloud_decimator.cpp

```diff
@@ -0,0 +1,174 @@
+      input_, 10, std::bind( &PointcloudDecimator::pointcloudCallback, this, std::placeholders::_1 ) );
+  RCLCPP_INFO( get_logger(), "Subscribed to '%s'", pointcloud_subscriber_->get_topic_name() );
+
+  // publisher for publishing outgoing messages
+  pointcloud_publisher_ = pct_->advertise( output_, 10 );
+  RCLCPP_INFO( get_logger(), "Publishing results to '%s'", pointcloud_publisher_.getTopic().c_str() );
+
+  check_subscribers_timer_ =
+      create_wall_timer( std::chrono::seconds( 1 ),
+                         std::bind( &PointcloudDecimator::publisherSubscriptionCallback, this ) );
+
+  std::srand( std::time( {} ) );
+}
+
+void PointcloudDecimator::pointcloudCallback( const sensor_msgs::msg::PointCloud2 &msg )
+{
+  const size_t point_step = msg.point_step;
+  const size_t input_size = msg.height * msg.width;
+
+  size_t point_count;
+  double point_fraction;
+  if ( elimination_quantifier_ == "count" ) {
+    point_count = std::min( static_cast<size_t>( point_count_ ), input_size );
+    point_fraction =
+        std::min( 1.0, static_cast<double>( point_count_ ) / static_cast<double>( input_size ) );
+  } else {
+    point_count = static_cast<size_t>( static_cast<double>( input_size ) * point_fraction_ );
+    point_fraction = point_fraction_;
+  }
+
+  RCLCPP_DEBUG( get_logger(), "Reducing %lu points to %lu", input_size, point_count );
+
+  // Copy shared fields
+  sensor_msgs::msg::PointCloud2 output;
+  output.header = msg.header;
+  output.height = 1;
+  output.fields = msg.fields;
+  output.is_bigendian = msg.is_bigendian;
+  output.point_step = msg.point_step;
+  output.is_dense = false;
+
+  output.data.clear();
+  output.data.reserve( point_count * point_step );
+
+  if ( elimination_method_ == "random" ) {
+    // Choose randomly which points to include. This will likely not match the chosen fraction/count exactly
+    for ( size_t point = 0; point < input_size; ++point ) {
+      const size_t source_index = point * point_step;
+
+      if ( static_cast<double>( std::rand() ) / RAND_MAX > point_fraction )
+        continue;
+
+      for ( size_t e = 0; e < point_step; ++e ) {
+        output.data.push_back( msg.data[source_index + e] );
+      }
+    }
+  } else {
+    for ( size_t point = 0; point < point_count; ++point ) {
+      const size_t source_index = static_cast<size_t>( point / point_fraction ) * point_step;
+
+      for ( size_t e = 0; e < point_step; ++e ) {
+        output.data.push_back( msg.data[source_index + e] );
+      }
+    }
+  }
+
+  output.row_step = output.data.size();
+  output.width = output.row_step / point_step;
+
+  pointcloud_publisher_.publish( output );
+}
+
+void PointcloudDecimator::publisherSubscriptionCallback()
+{
+
+  const size_t subscribers = pointcloud_publisher_.getNumSubscribers();
+
+  RCLCPP_DEBUG_STREAM( get_logger(), "Subscribers: " << subscribers
+                                                     << " has_subscribers_: " << has_subscribers_ );
```

Comment:
Please remove this debug stream. When increasing the check frequency, this will spam like crazy.